java/core/lance-jni/src/blocking_dataset.rs

```diff
@@ -812,3 +812,88 @@ fn inner_alter_columns(
     RT.block_on(dataset_guard.inner.alter_columns(&column_alterations))?;
     Ok(())
 }
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_Dataset_nativeAddColumnsBySqlExpressions(
+    mut env: JNIEnv,
+    java_dataset: JObject,
+    sql_expressions: JObject, // SqlExpressions
+    read_columns: JObject,    // Optional<List<String>>
+    batch_size: JObject,      // Optional<Long>
+) {
+    ok_or_throw_without_return!(
+        env,
+        inner_add_columns_by_sql_expressions(
+            &mut env,
+            java_dataset,
+            sql_expressions,
+            read_columns,
+            batch_size
+        )
+    )
+}
+
+fn inner_add_columns_by_sql_expressions(
+    env: &mut JNIEnv,
+    java_dataset: JObject,
+    sql_expressions: JObject, // SqlExpressions
+    read_columns: JObject,    // Optional<List<String>>
+    batch_size: JObject,      // Optional<Long>
+) -> Result<()> {
+    let sql_expressions_obj = env
+        .get_field(sql_expressions, "sqlExpressions", "Ljava/util/List;")?
+        .l()?;
+
+    let sql_expressions_obj_list = env.get_list(&sql_expressions_obj)?;
+    let mut expressions: Vec<(String, String)> = Vec::new();
+
+    for i in 0..sql_expressions_obj_list.size(env)? {
+        if let Ok(Some(item)) = sql_expressions_obj_list.get(env, i) {
```

Comment:
Why not use `.iter()` here? 
https://docs.rs/jni/latest/jni/objects/struct.JList.html#method.iter